[infrastructure] Refactor tests & extend coverage

jugglinmike opened the pull request:
testharness.js was recently extended with an API to explicitly opt-in to
the "single page test" feature [1]. As per WPT RFC 28 [2], tests which
do not use this API and which do not declare any subtests will soon be
reported as a harness error.

Update some tests which previously opted in implicitly to use the new
API. Update others to instead declare a single subtest (so that they are
no longer single-page tests). Add new tests to fully cover the
conditions under which an uncaught exception or unhandled rejection
could occur.

[1] https://github.com/web-platform-tests/wpt/pull/19449
[2] https://github.com/web-platform-tests/rfcs/blob/master/rfcs/single_test.md

jugglinmike committed:
[infrastructure] Refactor tests & extend coverage

testharness.js was recently extended with an API to explicitly opt-in to
the "single page test" feature [1]. As per WPT RFC 28 [2], tests which
do not use this API and which do not declare any subtests will soon be
reported as a harness error.

Update some tests which previously opted in implicitly to use the new
API. Update others to instead declare a single subtest (so that they are
no longer single-page tests). Add new tests to fully cover the
conditions under which an uncaught exception or unhandled rejection
could occur.

[1] https://github.com/web-platform-tests/wpt/pull/19449
[2] https://github.com/web-platform-tests/rfcs/blob/master/rfcs/single_test.md

foolip commented on infrastructure/browsers/firefox/prefs.html:
```diff
@@ -2,6 +2,7 @@
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <script>
+setup({ single_test: true });
```
Just wrap this one in `test(() => {})`

foolip commented on infrastructure/expected-fail/uncaught-exception.html:
```diff
@@ -4,5 +4,6 @@
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <script>
+setup({ single_test: true });
```
This and the following test was written to test when happens when there's a problem in unwrapped setup code. The expectations in infrastructure/metadata/ are what will have to change with the eventual change to remove implicit single-page tests.

jugglinmike replied:
Prior to the enforcement of `single_test`, there's no difference between "unwrapped setup code" and "failing single-page test code."

We'll need a new test for the "unwrapped setup code" when we begin enforcing `single_test`. Even then, we'll need to persist this test to preserve coverage for the "failing single-page test code" case.

foolip replied:
Having tests for both is fine, I'm just saying I want these to keep testing what they were, they're named appropriate for that.

jugglinmike replied:
The names "uncaught exception" and "unhandled rejection" don't describe the tests' effect on the harness. Until we enforce `single_test`, those events will be interpreted as test failures.

I can add tests for harness errors, but these will be new tests. In order to demonstrate this today, they will have to define a sub-test before producing the uncaught exception or unhandled rejection.

A new commit on this branch introduces those tests.

foolip replied:
> The names "uncaught exception" and "unhandled rejection" don't describe the tests' effect on the harness.

Indeed, I'm following @zcorpan's style of describing what the test does rather than its pass condition in the naming, since the pass condition may change, and will change here.

foolip replied:
Can you revert the change to infrastructure/expected-fail/uncaught-exception.html?

foolip commented on infrastructure/testdriver/file_upload.sub.html:
```diff
@@ -9,6 +9,7 @@
   <input id="file_input" name="file_input" type="file">
 </form>
 <script>
+setup({ single_test: true });
```
Wouldn't this one be better as a `promise_test`?

jugglinmike commented:
Since the commit message would no longer describe this solution, I've amended the proposed commit and force-pushed the result to this branch. [The original version of this branch is available here.](https://github.com/web-platform-tests/wpt/compare/master...bocoup:single-test-opt-in-infrastructure-orig)

foolip commented on infrastructure/expected-fail/unhandled-rejection.html:
```diff
@@ -4,5 +4,6 @@
 <script src="/resources/testharness.js"></script>
 <script src="/resources/testharnessreport.js"></script>
 <script>
+setup({ single_test: true });
```
I'd like this change to be reverted too. Another test that looks like this would be fine.

jugglinmike commented:
@foolip This patch now includes the tests you requested.

foolip commented:
Thanks @jugglinmike, looks great!

foolip commented:
@jugglinmike not sure exactly what commit message you'd like, so feel free to merge yourself.

jugglinmike commented:
Thanks, @foolip! I've squashed the commits, modified the commit message, force-pushed the result, and updated the pull request description accordingly. [The prior version of this branch is available here.](https://github.com/web-platform-tests/wpt/compare/master...bocoup:single-test-opt-in-infrastructure-orig-2)

Merging presently.
